### ISSUE TITLE
False positive "type alias is never used" when it's used in trait's generic argument

### ISSUE THREAD

**Issue by Riateche** [rust-lang/rust#67279]
Code (`lib.rs`):
```
mod a {
    trait X<A> {}
    type T = ();
    impl X<T> for () {}
}
```
`cargo check` produces `warning: type alias is never used: State1`.

**Comment by LeSeulArtichaut** [comment]
The lint originates from this function: https://github.com/rust-lang/rust/blob/0a58f5864659ddfe1d95c122abaa75c88220aed0/src/librustc_passes/dead.rs#L577
Which calls this other function: https://github.com/rust-lang/rust/blob/0a58f5864659ddfe1d95c122abaa75c88220aed0/src/librustc_passes/dead.rs#L494
Which calls yet another function: https://github.com/rust-lang/rust/blob/0a58f5864659ddfe1d95c122abaa75c88220aed0/src/librustc_passes/dead.rs#L527-L547

This is probably where code should change. I'll try to look at this, but I'm not sure I will be able to successfully implement the fix.

@rustbot claim

**Comment by Dylan-DPC** [comment]
This code compiles now without the warning
